bundles/org.eclipse.equinox.p2.core/src/org/eclipse/equinox/internal/p2/core/ProvisioningAgent.java

```diff
@@ -122,14 +109,10 @@ public void setLocation(URI location) {
 
 	@Override
 	public void unregisterService(String serviceName, Object service) {
-		synchronized (agentServices) {
-			if (stopped) {
-				return;
-			}
-			if (agentServices.get(serviceName) == service) {
-				agentServices.remove(serviceName);
-			}
+		if (stopped) {
+			return;
 		}
+		agentServices.remove(serviceName, service);
```

Comment:
Also for the `agentServices` I would use `ConcurrentHashMap`

Reply:
> Also for the `agentServices` I would use `ConcurrentHashMap`
`ConcurrentHashMap` is good for multiple concurrent reads, while here we deal with a short bursts of single-threaded ones. I think it would be slower and is not worth the additional memory footprint.